Feature/ms 147/enhanced readonly mode

anshul119 opened the pull request:
Ticket: https://oat-sa.atlassian.net/browse/MS-147

### What has changed?

* A new renderer(reviewRenderer) is added with interactions and choices in readonly/disabled mode
* A new helper function is added to map IMS view property to the right renderer
* All the choices and interactions which are not needed to be changed are inherited from qtiCommonRenderer
* Input fields and text areas are in `readonly` mode
* Select option fields, checkboxes and radio buttons are `disabled`
* All the click/tap and drag/drop event handlers are disabled with `pointer-events: none` style on templates

#### Related PRs:

- https://github.com/oat-sa/extension-tao-testqti-previewer/pull/63
- https://github.com/oat-sa/extension-tao-itemqti/pull/1439
- https://github.com/oat-sa/extension-lti-outcomeui/pull/13

### How to test?

- Checkout this branch in test previewer: `feature/MS-147/enhanced-readonly-mode`
- Checkout this branch in tao-item-qti: `feature/MS-147/enhanced-readonly-mode`
- Install node modules in tao-item-qti
- Create an item with any interactions. (I have an item with all interactions ready, you can ask me for that)
- Preview the item, you should see it as expected and MUST be able to interact with all interactions
- Take the test, you should see all items as usual and MUST be able to interact with all of them.
- Now, view the result of the test you submitted, now you are viewing the test with `reviewRenderer` so you won't be able to interact with any of the interactions as it is in read-only mode.


### Checklist:

- [ ] Merge this PR and publish package
- [ ] Change dependency version here and merge this PR: https://github.com/oat-sa/extension-tao-itemqti/pull/1439
- [ ] Merge this PR: https://github.com/oat-sa/extension-tao-testqti-previewer/pull/63
- [ ] Merge this PR: https://github.com/oat-sa/extension-lti-outcomeui/pull/13

jsconan commented on src/runner/provider/qti.js:
```diff
@@ -48,10 +49,15 @@ var qtiItemRuntimeProvider = {
         );
 
-        this._renderer = new QtiRenderer(rendererOptions);
+        if(rendererOptions.view === 'scorer') {
```
is this type accurate? should not it be something like `review` or `readonly` instead?

anshul119 replied:
According to IMS standards it is supposed to be called view with the following values:

- author
- candidate
- proctor
- scorer
- tutor 
- testConstructor

https://www.imsglobal.org/question/qtiv2p2p2/QTIv2p2p2-ASI-InformationModelv1p0/imsqtiv2p2p2_asi_v1p0_InfoModelv1p0.html#FigEnumeratedListClass_DataModel_View

btamas commented on src/reviewRenderer/renderers/config.js:
```diff
@@ -0,0 +1,129 @@
+    [
+        {
+            name: 'theme',
+            handle: function handleTheme(url) {
```
```suggestion
            handle(url) {
```

btamas commented on src/reviewRenderer/renderers/interactions/AssociateInteraction.js:
```diff
@@ -0,0 +1,37 @@
+ */
+import associateInteraction from 'taoQtiItem/qtiCommonRenderer/renderers/interactions/AssociateInteraction';
+
+var render = function(interaction) {
+    return new Promise(function(resolve) {
+        associateInteraction.renderEmptyPairs(interaction);
+
+        resolve();
+    });
+};
```
```suggestion
const render = interaction => Promise.resolve(associateInteraction.renderEmptyPairs(interaction));
```

btamas commented on src/reviewRenderer/renderers/interactions/ChoiceInteraction.js:
```diff
@@ -0,0 +1,40 @@
+import sizeAdapter from 'taoQtiItem/qtiCommonRenderer/helpers/sizeAdapter';
+
+var render = function render(interaction) {
+    var $container = containerHelper.get(interaction);
```
```suggestion
    const $container = containerHelper.get(interaction);
```

btamas commented on src/reviewRenderer/renderers/interactions/ChoiceInteraction.js:
```diff
@@ -0,0 +1,40 @@
+ * Expose the common renderer for the choice interaction
+ * @exports reviewRenderer/renderers/interactions/ChoiceInteraction
+ */
+export default Object.assign({}, choiceInteraction, {render: render});
```
```suggestion
export default Object.assign({}, choiceInteraction, { render });
```

btamas commented on src/reviewRenderer/renderers/interactions/ExtendedTextInteraction.js:
```diff
@@ -0,0 +1,44 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014 (original work) Open Assessment Technlogies SA (under the project TAO-PRODUCT);
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```

btamas commented on src/reviewRenderer/renderers/interactions/ExtendedTextInteraction.js:
```diff
@@ -0,0 +1,44 @@
+ * @param interaction
+ * @returns {*}
+ */
+var render = function render(interaction) {
```
```suggestion
const render = interaction => {
```

btamas commented on src/reviewRenderer/renderers/interactions/ExtendedTextInteraction.js:
```diff
@@ -0,0 +1,44 @@
+ * @exports qtiCommonRenderer/renderers/interactions/ExtendedTextInteraction
+ */
+
+export default Object.assign({}, extendedTextInteraction, {template: tpl, render: render});
```
```suggestion
export default Object.assign({}, extendedTextInteraction, { template, render });
```

btamas commented on src/reviewRenderer/renderers/interactions/ExtendedTextInteraction.js:
```diff
@@ -0,0 +1,44 @@
+/**
+ * @author Ansul Sharma <ansultaotesting.com>
+ */
+import tpl from 'taoQtiItem/reviewRenderer/tpl/interactions/extendedTextInteraction';
```
```suggestion
import template from 'taoQtiItem/reviewRenderer/tpl/interactions/extendedTextInteraction';
```

btamas commented on src/reviewRenderer/renderers/interactions/GapMatchInteraction.js:
```diff
@@ -0,0 +1,29 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014-2019 (original work) Open Assessment Technlogies SA (under the project TAO-PRODUCT);
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```

btamas commented on src/reviewRenderer/renderers/interactions/GraphicAssociateInteraction.js:
```diff
@@ -0,0 +1,30 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014 (original work) Open Assessment Technlogies SA (under the project TAO-PRODUCT);
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```

btamas commented on src/reviewRenderer/renderers/interactions/GraphicAssociateInteraction.js:
```diff
@@ -0,0 +1,30 @@
+/**
+ * @author Ansul Sharma <ansultaotesting.com>
+ */
+import tpl from 'taoQtiItem/reviewRenderer/tpl/interactions/graphicAssociateInteraction';
```
```suggestion
import template from 'taoQtiItem/reviewRenderer/tpl/interactions/graphicAssociateInteraction';
```

btamas commented on src/reviewRenderer/renderers/interactions/GraphicAssociateInteraction.js:
```diff
@@ -0,0 +1,30 @@
+ * Expose the common renderer for the hotspot interaction
+ * @exports qtiCommonRenderer/renderers/interactions/GraphicAssociateInteraction
+ */
+export default Object.assign({}, graphicAssociateInteraction, {template: tpl});
```
```suggestion
export default Object.assign({}, graphicAssociateInteraction, { template });
```

btamas commented on src/reviewRenderer/renderers/interactions/GraphicGapMatchInteraction.js:
```diff
@@ -0,0 +1,30 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014-2019 (original work) Open Assessment Technlogies SA (under the project TAO-PRODUCT);
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```

btamas commented on src/reviewRenderer/renderers/interactions/GraphicGapMatchInteraction.js:
```diff
@@ -0,0 +1,30 @@
+/**
+ * @author Ansul Sharma <ansultaotesting.com>
+ */
+import tpl from 'taoQtiItem/reviewRenderer/tpl/interactions/graphicGapMatchInteraction';
```
```suggestion
import template from 'taoQtiItem/reviewRenderer/tpl/interactions/graphicGapMatchInteraction';
```

btamas commented on src/reviewRenderer/renderers/interactions/GraphicGapMatchInteraction.js:
```diff
@@ -0,0 +1,30 @@
+ * Expose the common renderer for the hotspot interaction
+ * @exports qtiCommonRenderer/renderers/interactions/HotspotInteraction
+ */
+export default Object.assign({}, graphicGapMatchInteraction, {template: tpl});
```
```suggestion
export default Object.assign({}, graphicGapMatchInteraction, { template });
```

btamas commented on src/reviewRenderer/renderers/interactions/GraphicOrderInteraction.js:
```diff
@@ -0,0 +1,30 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014 (original work) Open Assessment Technlogies SA (under the project TAO-PRODUCT);
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```

btamas commented on src/reviewRenderer/renderers/interactions/GraphicOrderInteraction.js:
```diff
@@ -0,0 +1,30 @@
+/**
+ * @author Ansul Sharma <ansultaotesting.com>
+ */
+import tpl from 'taoQtiItem/reviewRenderer/tpl/interactions/graphicOrderInteraction';
```
```suggestion
import template from 'taoQtiItem/reviewRenderer/tpl/interactions/graphicOrderInteraction';
```

btamas commented on src/reviewRenderer/renderers/interactions/GraphicOrderInteraction.js:
```diff
@@ -0,0 +1,30 @@
+ * Expose the common renderer for the interaction
+ * @exports qtiCommonRenderer/renderers/interactions/SelectPointInteraction
+ */
+export default Object.assign({}, graphicOrderInteraction, {template: tpl});
```
```suggestion
export default Object.assign({}, graphicOrderInteraction, { template });
```

btamas commented on src/reviewRenderer/renderers/interactions/HotspotInteraction.js:
```diff
@@ -0,0 +1,30 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014 (original work) Open Assessment Technlogies SA (under the project TAO-PRODUCT);
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```

btamas commented on src/reviewRenderer/renderers/interactions/HotspotInteraction.js:
```diff
@@ -0,0 +1,30 @@
+/**
+ * @author Ansul Sharma <ansultaotesting.com>
+ */
+import tpl from 'taoQtiItem/reviewRenderer/tpl/interactions/hotspotInteraction';
```
```suggestion
import template from 'taoQtiItem/reviewRenderer/tpl/interactions/hotspotInteraction';
```

btamas commented on src/reviewRenderer/renderers/interactions/HotspotInteraction.js:
```diff
@@ -0,0 +1,30 @@
+ * Expose the common renderer for the hotspot interaction
+ * @exports qtiCommonRenderer/renderers/interactions/HotspotInteraction
+ */
+export default Object.assign({}, hotspotInteraction, {template: tpl});
```
```suggestion
export default Object.assign({}, hotspotInteraction, { template });
```

btamas commented on src/reviewRenderer/renderers/interactions/InlineChoiceInteraction.js:
```diff
@@ -0,0 +1,30 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014 (original work) Open Assessment Technlogies SA (under the project TAO-PRODUCT);
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```

btamas commented on src/reviewRenderer/renderers/interactions/InlineChoiceInteraction.js:
```diff
@@ -0,0 +1,30 @@
+/**
+ * @author Ansul Sharma <ansultaotesting.com>
+ */
+import tpl from 'taoQtiItem/reviewRenderer/tpl/interactions/inlineChoiceInteraction';
```
```suggestion
import template from 'taoQtiItem/reviewRenderer/tpl/interactions/inlineChoiceInteraction';
```

btamas commented on src/reviewRenderer/renderers/interactions/InlineChoiceInteraction.js:
```diff
@@ -0,0 +1,30 @@
+ * Expose the common renderer for the inline choice interaction
+ * @exports qtiCommonRenderer/renderers/interactions/InlineChoiceInteraction
+ */
+export default Object.assign({}, inlineChoiceInteraction, {template: tpl});
```
```suggestion
export default Object.assign({}, inlineChoiceInteraction, { template });
```

btamas commented on src/reviewRenderer/renderers/interactions/MatchInteraction.js:
```diff
@@ -0,0 +1,39 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014 (original work) Open Assessment Technlogies SA (under the project TAO-PRODUCT);
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```

btamas commented on src/reviewRenderer/renderers/interactions/MatchInteraction.js:
```diff
@@ -0,0 +1,39 @@
+import containerHelper from 'taoQtiItem/qtiCommonRenderer/helpers/container';
+import matchInteraction from 'taoQtiItem/qtiCommonRenderer/renderers/interactions/MatchInteraction';
+
+var render = function(interaction) {
```
```suggestion
const render =interaction => {
```

btamas commented on src/reviewRenderer/renderers/interactions/MatchInteraction.js:
```diff
@@ -0,0 +1,39 @@
+import matchInteraction from 'taoQtiItem/qtiCommonRenderer/renderers/interactions/MatchInteraction';
+
+var render = function(interaction) {
+    var $container = containerHelper.get(interaction);
```
```suggestion
    const $container = containerHelper.get(interaction);
```

btamas commented on src/reviewRenderer/renderers/interactions/MatchInteraction.js:
```diff
@@ -0,0 +1,39 @@
+var render = function(interaction) {
+    var $container = containerHelper.get(interaction);
+
+    $container.on('click.commonRenderer', 'input[type=checkbox]', function(e) {
+        e.preventDefault();
+    });
```
```suggestion
    $container.on('click.commonRenderer', 'input[type=checkbox]', e => e.preventDefault());
```

btamas commented on src/reviewRenderer/renderers/interactions/MatchInteraction.js:
```diff
@@ -0,0 +1,39 @@
+ * Expose the common renderer for the match interaction
+ * @exports reviewRenderer/renderers/interactions/MatchInteraction
+ */
+export default Object.assign({}, matchInteraction, {template: tpl, render: render});
```
```suggestion
export default Object.assign({}, matchInteraction, { template, render });
```

btamas commented on src/reviewRenderer/renderers/interactions/OrderInteraction.js:
```diff
@@ -0,0 +1,55 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014-2019 (original work) Open Assessment Technlogies SA (under the project TAO-PRODUCT);
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```

btamas commented on src/reviewRenderer/renderers/interactions/OrderInteraction.js:
```diff
@@ -0,0 +1,55 @@
+import orderInteraction from 'taoQtiItem/qtiCommonRenderer/renderers/interactions/OrderInteraction';
+import containerHelper from 'taoQtiItem/qtiCommonRenderer/helpers/container';
+
+var _freezeSize = function($container) {
+    var $orderArea = $container.find('.order-interaction-area');
```
```suggestion
const _freezeSize = $container => {
    const $orderArea = $container.find('.order-interaction-area');
```

btamas commented on src/reviewRenderer/renderers/interactions/OrderInteraction.js:
```diff
@@ -0,0 +1,55 @@
+ *
+ * @param {Object} interaction - the interaction instance
+ */
+var render = function(interaction) {
+    var $container = containerHelper.get(interaction);
```
```suggestion
const render = interaction => {
    const $container = containerHelper.get(interaction);
```

btamas commented on src/reviewRenderer/renderers/interactions/OrderInteraction.js:
```diff
@@ -0,0 +1,55 @@
+    var $container = containerHelper.get(interaction);
+
+    //bind event listener in case the attributes change dynamically on runtime
+    $(document).on('attributeChange.qti-widget.commonRenderer', function(e) {
+        e.preventDefault();
+    });
```
```suggestion
    $(document).on('attributeChange.qti-widget.commonRenderer', e => e.preventDefault());
```

btamas commented on src/reviewRenderer/renderers/interactions/OrderInteraction.js:
```diff
@@ -0,0 +1,55 @@
+ * Expose the common renderer for the order interaction
+ * @exports qtiCommonRenderer/renderers/interactions/OrderInteraction
+ */
+export default Object.assign({}, orderInteraction, {render: render});
```
```suggestion
export default Object.assign({}, orderInteraction, { render });
```

btamas commented on src/reviewRenderer/renderers/interactions/PortableCustomInteraction.js:
```diff
@@ -0,0 +1,26 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014-2017 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```

btamas commented on src/reviewRenderer/renderers/interactions/PortableCustomInteraction.js:
```diff
@@ -0,0 +1,26 @@
+/**
+ * @author Ansul Sharma <ansultaotesting.com>
+ */
+import tpl from 'taoQtiItem/reviewRenderer/tpl/interactions/customInteraction';
```
```suggestion
import template from 'taoQtiItem/reviewRenderer/tpl/interactions/customInteraction';
```

btamas commented on src/reviewRenderer/renderers/interactions/PortableCustomInteraction.js:
```diff
@@ -0,0 +1,26 @@
+import tpl from 'taoQtiItem/reviewRenderer/tpl/interactions/customInteraction';
+import portableCustomInteraction from 'taoQtiItem/qtiCommonRenderer/renderers/interactions/PortableCustomInteraction';
+
+export default Object.assign({}, portableCustomInteraction, {template: tpl});
```
```suggestion
export default Object.assign({}, portableCustomInteraction, { template });
```

btamas commented on src/reviewRenderer/renderers/interactions/SelectPointInteraction.js:
```diff
@@ -0,0 +1,30 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014-2017 (original work) Open Assessment Technlogies SA (under the project TAO-PRODUCT);
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```

btamas commented on src/reviewRenderer/renderers/interactions/SelectPointInteraction.js:
```diff
@@ -0,0 +1,30 @@
+/**
+ * @author Ansul Sharma <ansultaotesting.com>
+ */
+import tpl from 'taoQtiItem/reviewRenderer/tpl/interactions/selectPointInteraction';
```
```suggestion
import template from 'taoQtiItem/reviewRenderer/tpl/interactions/selectPointInteraction';
```

btamas commented on src/reviewRenderer/renderers/interactions/SelectPointInteraction.js:
```diff
@@ -0,0 +1,30 @@
+ * Expose the common renderer for the interaction
+ * @exports qtiCommonRenderer/renderers/interactions/SelectPointInteraction
+ */
+export default Object.assign({}, selectPointInteraction, {template: tpl});
```
```suggestion
export default Object.assign({}, selectPointInteraction, { template });
```

btamas commented on src/reviewRenderer/renderers/interactions/SliderInteraction.js:
```diff
@@ -0,0 +1,26 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014 (original work) Open Assessment Technlogies SA (under the project TAO-PRODUCT);
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```

btamas commented on src/reviewRenderer/renderers/interactions/SliderInteraction.js:
```diff
@@ -0,0 +1,26 @@
+/**
+ * @author Ansul Sharma <ansultaotesting.com>
+ */
+import tpl from 'taoQtiItem/reviewRenderer/tpl/interactions/sliderInteraction';
```
```suggestion
import template from 'taoQtiItem/reviewRenderer/tpl/interactions/sliderInteraction';
```

btamas commented on src/reviewRenderer/renderers/interactions/SliderInteraction.js:
```diff
@@ -0,0 +1,26 @@
+import tpl from 'taoQtiItem/reviewRenderer/tpl/interactions/sliderInteraction';
+import sliderInteraction from 'taoQtiItem/qtiCommonRenderer/renderers/interactions/SliderInteraction';
+
+export default Object.assign({}, sliderInteraction, {template: tpl});
```
```suggestion
export default Object.assign({}, sliderInteraction, { template });
```

btamas commented on src/reviewRenderer/renderers/interactions/TextEntryInteraction.js:
```diff
@@ -0,0 +1,26 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```

btamas commented on src/reviewRenderer/renderers/interactions/TextEntryInteraction.js:
```diff
@@ -0,0 +1,26 @@
+/**
+ * @author Ansul Sharma <ansultaotesting.com>
+ */
+import tpl from 'taoQtiItem/reviewRenderer/tpl/interactions/textEntryInteraction';
```
```suggestion
import template from 'taoQtiItem/reviewRenderer/tpl/interactions/textEntryInteraction';
```

btamas commented on src/reviewRenderer/renderers/interactions/TextEntryInteraction.js:
```diff
@@ -0,0 +1,26 @@
+import tpl from 'taoQtiItem/reviewRenderer/tpl/interactions/textEntryInteraction';
+import textEntryInteraction from 'taoQtiItem/qtiCommonRenderer/renderers/interactions/TextEntryInteraction';
+
+export default Object.assign({}, textEntryInteraction, {template: tpl});
```
```suggestion
export default Object.assign({}, textEntryInteraction, { template });
```

btamas commented on src/reviewRenderer/renderers/interactions/UploadInteraction.js:
```diff
@@ -0,0 +1,26 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014 (original work) Open Assessment Technlogies SA (under the project TAO-PRODUCT);
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```

btamas commented on src/reviewRenderer/renderers/interactions/UploadInteraction.js:
```diff
@@ -0,0 +1,26 @@
+/**
+ * @author Ansul Sharma <ansultaotesting.com>
+ */
+import tpl from 'taoQtiItem/reviewRenderer/tpl/interactions/uploadInteraction';
```
```suggestion
import template from 'taoQtiItem/reviewRenderer/tpl/interactions/uploadInteraction';
```

btamas commented on src/reviewRenderer/renderers/interactions/UploadInteraction.js:
```diff
@@ -0,0 +1,26 @@
+import tpl from 'taoQtiItem/reviewRenderer/tpl/interactions/uploadInteraction';
+import uploadInteraction from 'taoQtiItem/qtiCommonRenderer/renderers/interactions/UploadInteraction';
+
+export default Object.assign({}, uploadInteraction, {template: tpl});
```
```suggestion
export default Object.assign({}, uploadInteraction, { template });
```

btamas commented on src/reviewRenderer/renderers/rendererProvider.js:
```diff
@@ -0,0 +1,27 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014 (original work) Open Assessment Technologies SA;
```
```suggestion
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
```

btamas commented on src/reviewRenderer/renderers/rendererProvider.js:
```diff
@@ -0,0 +1,27 @@
+
+export default {
+    name: 'reviewRenderer',
+    init: function() {},
+    getRenderer: function() {
+        return ReviewRenderer;
+    }
```
```suggestion
    init: () => {},
    getRenderer: () => ReviewRenderer
```

btamas commented on src/runner/provider/qti.js:
```diff
@@ -29,10 +29,11 @@ import Element from 'taoQtiItem/qtiItem/core/Element';
 import icRegistry from 'taoQtiItem/portableElementRegistry/icRegistry';
 import sideLoadingProviderFactory from 'taoQtiItem/portableElementRegistry/provider/sideLoadingProviderFactory';
-import QtiRenderer from 'taoQtiItem/qtiCommonRenderer/renderers/Renderer';
+import rendererManager from 'taoQtiItem/runner/rendererManager';
```
License header should be updated.

btamas commented on src/runner/rendererManager.js:
```diff
@@ -0,0 +1,23 @@
+import providerRegistry from 'core/providerRegistry';
```
Missing license header.

btamas commented on src/runner/utils/imsViewToRendererMaper.js:
```diff
@@ -0,0 +1,13 @@
+const getRendererNameFromView = function(imsView) {
```
Missing license header.

btamas commented on src/runner/utils/imsViewToRendererMaper.js:
```diff
@@ -0,0 +1,13 @@
+const getRendererNameFromView = function(imsView) {
+    let renderer = '';
+
+    if (imsView === 'scorer') {
+        renderer = 'reviewRenderer';
+    } else {
+        renderer = 'commonRenderer';
+    }
+
+    return renderer;
```
```suggestion
    return imsView === 'scorer' ? 'reviewRenderer' : 'commonRenderer';
```

btamas commented on src/runner/provider/qti.js:
```diff
@@ -13,7 +13,7 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  *
- * Copyright (c) 2014 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
+ * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```
If it is not a new file you should set it to `2014-2020`

krampstudio commented:
It's a good idea to switch between renderers implementations. So it goes into the right direction. 
It just wasn't so obvious with the name "rendererManager" and each provider is only a link to the renderer (`getRenderer`). 
Why don't you apply the delegation (provider pattern) directly to the `Renderer` itsself ?

anshul119 commented:
> It's a good idea to switch between renderers implementations. So it goes into the right direction.
> It just wasn't so obvious with the name "rendererManager" and each provider is only a link to the renderer (`getRenderer`).
> Why don't you apply the delegation (provider pattern) directly to the `Renderer` itsself ?

1. Maybe `rendererRegistry` is a better name, what do you think?
2. I didn't want to break the code in current places which are importing the Renderer.js file. Eg: qtiPrint, qtiCreator, qtiXmlRenderer, etc. Hence, I created a provider wrapper around Renderer. So the existing extensions can keep using the commonRenderer directly. Do you suggest o do this another way?

krampstudio commented:
> > It's a good idea to switch between renderers implementations. So it goes into the right direction.
> > It just wasn't so obvious with the name "rendererManager" and each provider is only a link to the renderer (`getRenderer`).
> > Why don't you apply the delegation (provider pattern) directly to the `Renderer` itsself ?
> 
>     1. Maybe `rendererRegistry` is a better name, what do you think?
> 
>     2. I didn't want to break the code in current places which are importing the Renderer.js file. Eg: qtiPrint, qtiCreator, qtiXmlRenderer, etc. Hence, I created a provider wrapper around Renderer. So the existing extensions can keep using the commonRenderer directly. Do you suggest o do this another way?

It's up to you, it depends how safe it is to change it. Otherwise you can keep this way with a better name and documentation

andrey-ponamarev commented on src/runner/utils/imsViewToRendererMaper.js:
```diff
@@ -17,6 +17,16 @@
+ * Returns the appropriate renderer based on the IMS view property.
+ *
+ * Read more about IMS view here:
+ * https://www.imsglobal.org/question/qtiv2p2p2/QTIv2p2p2-ASI-InformationModelv1p0/imsqtiv2p2p2_asi_v1p0_InfoModelv1p0.html#FigEnumeratedListClass_DataModel_View
```
👍

anshul119 commented:
> > > It's a good idea to switch between renderers implementations. So it goes into the right direction.
> > > It just wasn't so obvious with the name "rendererManager" and each provider is only a link to the renderer (`getRenderer`).
> > > Why don't you apply the delegation (provider pattern) directly to the `Renderer` itsself ?
> > 
> > 
> > ```
> > 1. Maybe `rendererRegistry` is a better name, what do you think?
> > 
> > 2. I didn't want to break the code in current places which are importing the Renderer.js file. Eg: qtiPrint, qtiCreator, qtiXmlRenderer, etc. Hence, I created a provider wrapper around Renderer. So the existing extensions can keep using the commonRenderer directly. Do you suggest o do this another way?
> > ```
> 
> It's up to you, it depends how safe it is to change it. Otherwise you can keep this way with a better name and documentation

I have added some extra comments and changed the name to `rendererStrategies`
